Manages cross-distribution dependencies better (#58430)
mistercrunch/Airflow: contributing-docs/13_airflow_dependencies_and_extras.rst

Airflow dependencies
====================

This document describes how we manage Apache Airflow dependencies, as well as how we make sure
that users can use Airflow both as an application and as a library when they are deploying
their own Airflow - using our constraints mechanism.

Airflow ``pyproject.toml`` files and ``uv`` workspace
.....................................................

> Managing dependencies is an important part of maintaining Apache Airflow, we have more than 700

Reviewer: ```suggestion
Managing dependencies is an important part of developing Apache Airflow, we have more than 700
```

To me, maintaining Airflow is things you do with a running Airflow deployment.

Author: Fine for me :). Indeed maintainers might do a lot more with dependencies than developers and this guideline is mostly for developers.

Adding and modifying dependencies
.................................

Adding and modifying dependencies in Apache Airflow is done by modifying the appropriate
``pyproject.toml`` file in the appropriate distribution.

When you add a new dependency, you should make sure that:

* The dependency is added to the right section (main dependencies, optional dependencies, or
  development dependencies)

* Some parts of those dependencies might be automatically generated (and overwritten) by our ``prek``
  hooks. Those are the necessary dependencies that ``prek`` hoks can figure out automatically by
  analyzing the imports in the sources and structure of the project. We also have special case of
  shared dependencies (described in `shared dependencies document <../shared/README.md>`__) where we
  effectively "static-link" some libraries into multiple distributions, to avoid unnecessary coupling and
  circular dependencies, and those dependencies contribute some dependencies automatically as well.
  Pay attention to comments such us example start and end of such generated block below.
  All the dependencies between those comments are automatically generated and you should not modify

>   them manually. Tou might instead modify the root source of those dependencies - depending on the automation,

Reviewer: ```suggestion
  them manually. You might instead modify the root source of those dependencies - depending on the automation,
```

  conflicts between dependencies of different packages in the workspace. You can run it in multiple
  ways - either from the root of the repository (which will sync all packages) or from the package
  you modified (which will sync only that package and its dependencies). Also good idea might be to
  run ``uv sync --all-packages --all-extras`` at the root of the repository to make sure that
  all packages with all extras can be installed together without conflicts, but this might be sometimes
  difficult and slow as some of the extras require some additional system level dependencies to be installed
  (for example ``mysql`` or ``postgres`` extras require client libraries to be installed on the system).

* Make sure to run all tests after modifying dependencies to make sure that nothing is broken

Referring to other Apache Airflow distributions in dependencies
...............................................................

With having more than 100 distributions in the repository, it is often necessary to refer to
other distributions in the dependencies in order to use some common features or simply to use the

> features that the other distribution provides. There are two ways of doing it:

Reviewer: This is the third time we mention how many dists we have.

```suggestion
It is often necessary to refer to other distributions in the dependencies. There are two
ways of doing it:
```

Author: Again - repetitions are kind of expected and built in. 

I 100% do **NOT** expect people to read that as a book. Or even actually read it from the beginning to an end even once. Not even close.

I do expect people to look for particular issue they have and find the decription about it without them going back to the beginning of the doc and reading it from the very beginning (but I do expect AI agents reading all of it and making it part of their context) - so sometimss repetitions are very deliberate in this docs. 

But yeah - especially that we might have 200+ soon, maybe repeating the number itself makes no sense.

  creating circular dependencies.

We are not going to describe the shared dependencies mechanism here, please refer to the
`shared dependencies document <../shared/README.md>`__ for details, but there are certain rules
when it comes to referring to other Airflow distributions in dependencies - here are the important
rules to remember:

* You can refer to other distributions in your dependencies - as usual using distribution name. For example,
  if you are adding a dependency to ``apache-airflow-providers-common-compat`` package from
  ``apache-airflow-providers-google``, you can just add ``apache-airflow-providers-common>=x.y.z`` to the
  dependencies and when you run ``uv sync``, the local version of the package will be used automatically
  (this is thanks to the workspace feature of ``uv`` that does great job of binding our monorepo together).
  This is very useful when you are developing changes to multiple packages at the same time. Some of those
  are added automatically by prek hooks - when it can detect such dependencies by analyzing imports in the
  sources - then they are added automatically between the special comments mentioned above, but sometimes

>   (especially when such dependencies are not at the top-level imports) you might need to add them manually.

Reviewer: You have already mentioned the developing changes locally above. Adding it just makes the point harder to follow.

Author: Yeah. Taht was too much.

* Never update distribution versions on your own. It is **entirely** up to the Release Manager to bump the
  version of distributions that are defined as ``project.version``. This goes almost without an exception
  and any diversions from this rule should be discussed at ``#release-management`` channel in Airflow Slack
  beforehand. The only exception to this rule is when you are adding a new distribution to the repository -
  in that case you should set the initial version of the distribution  - usually ``0.0.1``, ``0.1.0``
  or ``1.0.0`` depending on the maturity of the package. But still you should discuss it in the channel.

> 

Reviewer: This feels out of place in a document talking about dependencies.

```suggestion
```

Author: This is actually part of it and it makes sense in terms of internal cross-dependencies - and came up in maaaany conversations:

"I added this feature, should I update the version of this distribution or will RM do it for me". People were absolutely not clear about it.

But I added this sentence before to make it clearer:

> In case you make a feature change in a distribution and would like to update its version, you should never update the distribution version on your own

* Sometimes, when you add a new feature to a common distribution, you might add a feature to it or
  change the API in the way that other packages can use it. This is especially true for common packages such as
  ``apache-airflow-providers-common-compat``, but can happen for other packages (for example
  ``apache-airflow-providers-apache-beam`` is used by ``apache-airflow-providers-google`` to use ``Apache Beam``
  hooks to communicate with Google Dataflow). In such case, when you are adding a feature to a common package

>   remember that the feature you just add. will only be released in the **FUTURE** release of such common

Reviewer: ```suggestion
  remember that the feature you just add will only be released in the **FUTURE** release of such common
```

  package and you cannot add ``>==x.y.z`` dependency to it where ``x.y.z`` is the version you are
  going to release in the future. Ultimately, this should happen (and happens) when the Release Manager prepares
  both packages together. Let us repeat - such changes in versions between different airflow package should
  **NOT** be added to the dependencies manually by the contributor. They should **exclusively** be added by
  the Release Manager. when preparing the release of **both** packages together.
  We have a custom mechanism to support such additions, where it is contributor's responsibility to mark
  dependency with a special comment - simply communicating with the Release Manager that such dependency
  should be updated to the next version when the release is prepared. If you see such a need to use newly
  added feature and using it at the same time in a different distribution -  make sure to add this comment
  in the line where dependency you want to use the new feature from is defined:

  .. code-block:: python

>       # some deps

Reviewer: At first I thought "some deps" was an important comment.
```suggestion
      # ...
      # other deps here
      # ...
```

Author: Ah yeah. Fixed.

>       "apache-airflow-SOMETHING>1.0.0",  # use next version

Reviewer: Is `use next version` an instruction to a human, or a precise string that CI picks up and must match?

Author: It is precise string. I will make it clearer.

Author: Good point.
